File: src/DAO/CategoriaDAO.java
package DAO;

import CX.ConexionBD;
import Modelo.Categoria;
import Utils.AESUtil;
import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class CategoriaDAO {

    public List<Categoria> listarTodas() throws SQLException {
        List<Categoria> lista = new ArrayList<>();
        String sql = "SELECT * FROM categorias ORDER BY nombre";
        try (Connection con = ConexionBD.conectar();
             PreparedStatement ps = con.prepareStatement(sql);
             ResultSet rs = ps.executeQuery()) {
            while (rs.next()) lista.add(mapear(rs));
        }
        return lista;
    }

    public void insertar(Categoria c) throws SQLException {
        String sql = "INSERT INTO categorias (nombre, descripcion) VALUES (?, ?)";
        try (Connection con = ConexionBD.conectar();
             PreparedStatement ps = con.prepareStatement(sql)) {
            ps.setString(1, AESUtil.encriptar(c.getNombre()));
            ps.setString(2, AESUtil.encriptar(c.getDescripcion()));
            ps.executeUpdate();
        }
    }

    public void actualizar(Categoria c) throws SQLException {
        String sql = "UPDATE categorias SET nombre=?, descripcion=? WHERE id=?";
        try (Connection con = ConexionBD.conectar();
             PreparedStatement ps = con.prepareStatement(sql)) {
            ps.setString(1, AESUtil.encriptar(c.getNombre()));
            ps.setString(2, AESUtil.encriptar(c.getDescripcion()));
            ps.setLong(3, c.getId());
            ps.executeUpdate();
        }
    }

    public void eliminar(long id) throws SQLException {
        String sql = "DELETE FROM categorias WHERE id=?";
        try (Connection con = ConexionBD.conectar();
             PreparedStatement ps = con.prepareStatement(sql)) {
            ps.setLong(1, id);
            ps.executeUpdate();
        }
    }

    private Categoria mapear(ResultSet rs) throws SQLException {
        return new Categoria(
            rs.getLong("id"), 
            AESUtil.desencriptar(rs.getString("nombre")), 
            AESUtil.desencriptar(rs.getString("descripcion"))
        );
    }
}

File: inventario_python/dao/categoria_dao.py
from db.conexion import conectar
from models.categoria import Categoria
from utils.aes_util import encriptar, desencriptar


class CategoriaDAO:

    def _mapear(self, row):
        c = Categoria()
        c.id = row[0]
        c.nombre = desencriptar(row[1]) if row[1] else ""
        c.descripcion = desencriptar(row[2]) if row[2] else ""
        return c

    def listar_todas(self):
        conn = conectar()
        if not conn:
            return []
        try:
            cur = conn.cursor()
            cur.execute("SELECT * FROM categorias ORDER BY id DESC")
            return [self._mapear(r) for r in cur.fetchall()]
        finally:
            conn.close()

    def insertar(self, c):
        conn = conectar()
        if not conn:
            return False
        try:
            cur = conn.cursor()
            cur.execute("INSERT INTO categorias (nombre, descripcion) VALUES (%s, %s)",
                        (encriptar(c.nombre), encriptar(c.descripcion) if c.descripcion else None))
            conn.commit()
            return True
        except Exception as e:
            conn.rollback()
            raise e
        finally:
            conn.close()

    def actualizar(self, c):
        conn = conectar()
        if not conn:
            return False
        try:
            cur = conn.cursor()
            cur.execute("UPDATE categorias SET nombre=%s, descripcion=%s WHERE id=%s",
                        (encriptar(c.nombre), encriptar(c.descripcion) if c.descripcion else None, c.id))
            conn.commit()
            return True
        except Exception as e:
            conn.rollback()
            raise e
        finally:
            conn.close()

    def eliminar(self, id):
        conn = conectar()
        if not conn:
            return False
        try:
            cur = conn.cursor()
            cur.execute("DELETE FROM categorias WHERE id = %s", (id,))
            conn.commit()
            return True
        except Exception as e:
            conn.rollback()
            raise e
        finally:
            conn.close()
